### PR TITLE
Soft discrete byol

### PR DESCRIPTION
**Pull request by lavoiems**

### CODE REVIEW

File: bash_files/pretrain/cifar/sdbyol.sh [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    --contrast 0.4 \
+    --saturation 0.2 \
+    --hue 0.1 \
+    --gaussian_prob 0.0 0.0 \
+    --solarization_prob 0.0 0.2 \
+    --crop_size 32 \
+    --num_crops_per_aug 1 1 \
+    --name sdbyol-$1 \
+    --wandb \
+    --project solo-learn \
```
Comment (tsirif) [inline bash_files/pretrain/cifar/sdbyol.sh]: let's put these to out wandb project and entity, in case you have forgotten

File: bash_files/pretrain/cifar/sdbyol.sh [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    --optimizer sgd \
+    --lars \
+    --grad_clip_lars \
+    --eta_lars 0.02 \
+    --exclude_bias_n_norm \
+    --scheduler warmup_cosine \
+    --lr 1.0 \
+    --classifier_lr 0.03 \
+    --weight_decay 1e-6 \
+    --batch_size 256 \
```
Comment (tsirif) [inline bash_files/pretrain/cifar/sdbyol.sh:19]: I am guessing this fits in a single a100 right?

File: bash_files/pretrain/imagenet/sdbyol.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
+
+ROOT_PATH=$1
+
+../../../prepare_data.sh VAL
```
Comment (tsirif) [inline bash_files/pretrain/imagenet/sdbyol.sh:5]: this is the script that takes imagenet at mila and splits it to train/val/test right?

File: bash_files/pretrain/imagenet/sdbyol.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+python3 ../../../main_pretrain.py \
+    --dataset imagenet \
+    --backbone resnet50 \
+    --checkpoint_dir=$ROOT_PATH \
+    --data_dir $SLURM_TMPDIR/data \
+    --train_dir $SLURM_TMPDIR/data/train/ \
+    --val_dir   $SLURM_TMPDIR/data/val/ \
+    --max_epochs 100 \
+    --gpus 0,1 \
```
Comment (tsirif) [inline bash_files/pretrain/imagenet/sdbyol.sh:18]: does this mean "use 2 gpus with id 0 and 1"?

File: bash_files/pretrain/imagenet/sdbyol.sh (hunk 3 above)
Comment (tsirif) [reply bash_files/pretrain/imagenet/sdbyol.sh:5]: Is this script deterministic? will it provide with the same splits every time?

File: solo/methods/sdbyol.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+            nn.BatchNorm1d(proj_hidden_dim),
+            nn.ReLU(),
+            nn.Linear(proj_hidden_dim, proj_output_dim),
+        )
+
+        # Momentum
+        self.momentum_embedder = nn.Sequential(
+                            nn.Linear(self.features_dim, message_size*voc_size, bias=False),
+                            nn.BatchNorm1d(message_size*voc_size))
+        self.momentum_softmax = SoftmaxBridge(message_size, voc_size, tau_target, **kwargs)
```
Comment (tsirif) [inline solo/methods/sdbyol.py:100]: would have been the same as:
`self.momentum_embedder = copy.deepcopy(self.embedder)`
`self.momentum_embedder.requires_grad_(False)`

File: solo/methods/sdbyol.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+        """
+
+        extra_learnable_params = [
+            {"params": self.projector.parameters()},
+            {"params": self.predictor.parameters()},
+            {"params": self.embedder.parameters()},
+            {"name": 'classifier_y',
+              "params": self.linear_y.parameters(),
+              "lr": self.classifier_lr,
+              "weight_decay": 0
```
Comment (tsirif) [inline solo/methods/sdbyol.py:159]: no ridge here for the linear classifier on y?

File: solo/methods/sdbyol.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+        Returns:
+            Dict[str, Any]: a dict containing the outputs of the parent and the projected features.
+        """
+
+        out = super().forward(X, *args, **kwargs)
+        emb = self.embedder(out['feats'])
+        y = self.softmax(emb)
+        z = self.projector(y)
+        p = self.predictor(z)
+        return {**out, "z": z, "p": p, "logits": logits, "y": y}
```
Comment (tsirif) [inline solo/methods/sdbyol.py]: undefined name `logits` here

File: solo/methods/sdbyol.py [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+        return {"loss": loss, "acc1": acc1, "acc5": acc5}
+
+    def _shared_step(
+        self, feats: List[torch.Tensor], momentum_feats: List[torch.Tensor], targets: List[torch.tensor]
+    ) -> torch.Tensor:
+
+        embs = [self.embedder(f) for f in feats]
+        Y = [self.softmax(e) for e in embs]
+        Z = [self.projector(y) for y in Y]
+        P = [self.predictor(z) for z in Z]
```
Comment (tsirif) [inline solo/methods/sdbyol.py:206]: I don't know why we are doing it like this here... is it for accumulating mini-batches?

File: solo/methods/sdbyol.py [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+        with torch.no_grad():
+            embs_momentum = [self.momentum_embedder(f) for f in momentum_feats]
+            Y_momentum = [self.momentum_softmax(e) for e in embs_momentum]
+            Z_momentum = [self.momentum_projector(y) for y in Y_momentum]
+
+        # ------- negative consine similarity loss -------
+        neg_cos_sim = 0
+        for v1 in range(self.num_large_crops):
+            for v2 in np.delete(range(self.num_crops), v1):
+                neg_cos_sim += byol_loss_func(P[v2], Z_momentum[v1])
```
Comment (tsirif) [inline solo/methods/sdbyol.py:218]: does `+=` work now with pytorch and autodiff? I thought we had to write always `neg_cos_sim = neg_cos_sim + ...`

Comment (tsirif) [reply solo/methods/sdbyol.py:218]: is `self.num_crops` the same as `self.num_large_crops` here?
so the list is for different views??

File: solo/methods/sdbyol.py [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+
+        # calculate std of features
+        with torch.no_grad():
+            z_std = F.normalize(torch.stack(Z[: self.num_large_crops]), dim=-1).std(dim=1).mean()
```
Comment (tsirif) [inline solo/methods/sdbyol.py:222]: is this for debugging?

File: solo/methods/sdbyol.py [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+        metrics = {
+            "train_neg_cos_sim": neg_cos_sim,
+            "train_z_std": z_std,
+        }
+
+        metrics.update(online_class)
+
+        self.log_dict(metrics, on_epoch=True, sync_dist=True)
+
+        return neg_cos_sim + class_loss + online_class_loss
```
Comment (tsirif) [inline solo/methods/sdbyol.py:267]: gg!

File: bash_files/pretrain/cifar/sdbyol.sh (hunk 1 above)
Comment (lavoiems) [reply bash_files/pretrain/cifar/sdbyol.sh]: Good idea

File: bash_files/pretrain/cifar/sdbyol.sh (hunk 2 above)
Comment (lavoiems) [reply bash_files/pretrain/cifar/sdbyol.sh:19]: It does

File: bash_files/pretrain/imagenet/sdbyol.sh (hunk 3 above)
Comment (lavoiems) [reply bash_files/pretrain/imagenet/sdbyol.sh:5]: Yes, it takes the 10 first images in each directory.

File: bash_files/pretrain/imagenet/sdbyol.sh (hunk 4 above)
Comment (lavoiems) [reply bash_files/pretrain/imagenet/sdbyol.sh:18]: Yes

File: solo/methods/sdbyol.py (hunk 6 above)
Comment (lavoiems) [reply solo/methods/sdbyol.py:159]: No, otherwise it would be too slow for imagenet.

File: solo/methods/sdbyol.py (hunk 7 above)
Comment (lavoiems) [reply solo/methods/sdbyol.py]: Good catch. I'll fix it.

File: solo/methods/sdbyol.py (hunk 10 above)
Comment (lavoiems) [reply solo/methods/sdbyol.py:222]: Yea, they had this in byol.py.

File: solo/methods/sdbyol.py (hunk 8 above)
Comment (lavoiems) [reply solo/methods/sdbyol.py:206]: It is accumulating the views to be used when computing the loss.

File: solo/methods/sdbyol.py (hunk 9 above)
Comment (lavoiems) [reply solo/methods/sdbyol.py:218]: It does work as far as I can tell, it does train. However, maybe I am missing something.

Yes, num_crops is generally the same as num_large_crops. But, I guess that they could be different.

This for loop is a trick that they implemented for handling multi-cropping. Maybe not the sexiest implementation :P

File: solo/methods/sdbyol.py (hunk 11 above)
Comment (lavoiems) [reply solo/methods/sdbyol.py:267]: :) Thanks for the good review

File: solo/methods/sdbyol.py (hunk 5 above)
Comment (lavoiems) [reply solo/methods/sdbyol.py:100]: Yea, L.109 `initialize_momentum_params(self.embedder, self.momentum_embedder)` does essentially that.
